### PR TITLE
Add Volume Screen test and Haptics

### PR DESCRIPTION
**Pull request by kul3r4**

### CODE REVIEW

File: audio-ui/src/main/java/com/google/android/horologist/audioui/devices/Devices.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import com.google.android.horologist.audioui.devices.PremiumHaptics
+import com.google.android.horologist.audioui.devices.PrimitiveHaptics
+import com.google.android.horologist.audioui.devices.RotaryInput
+
+public val LocalHaptics: ProvidableCompositionLocal<Haptics> =
+    staticCompositionLocalOf { Haptics.None }
+public val LocalRotaryInput: ProvidableCompositionLocal<RotaryInput?> =
+    staticCompositionLocalOf { null }
+
+public fun haptics(context: Context): Haptics {
```
Comment (yschimke) [inline audio-ui/src/main/java/com/google/android/horologist/audioui/devices/Devices.kt:35]: I'm not sure we are ready for all of this, I think it needs more discussion.  Could we remove most of this from this PR, maybe just keep Haptics interface and one default implementation, and rely on caller to pass this in for now?

File: audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt [hunk 2]
```diff
@@ -64,4 +75,76 @@ public class VolumeScrollableState(private val volumeRepository: VolumeRepositor
     ) {
         scrollableState.scroll(block = block)
     }
+
+    private fun performHaptics(){
+        if (VERSION.SDK_INT >= VERSION_CODES.R) {
+            var effect: Int
+            for (primitive in COMPOSITION_PRIMITIVES) {
```
Comment (yschimke) [inline audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt:82]: Why performing all 4 here? I would have expected a single haptic event here, tick or click?

File: audio-ui/src/androidTest/java/com/google/android/horologist/audioui/TestHaptics.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import com.google.android.horologist.audioui.devices.Haptics
```
Comment (yschimke) [inline audio-ui/src/androidTest/java/com/google/android/horologist/audioui/TestHaptics.kt:1]: Missing package statement, and import should be below copyright.

File: audio-ui/src/main/java/com/google/android/horologist/audioui/devices/Devices.kt (hunk 1 above)
Comment (yschimke) [reply audio-ui/src/main/java/com/google/android/horologist/audioui/devices/Devices.kt:35]: This probably applies to the whole devices package.

File: audio-ui/src/androidTest/java/com/google/android/horologist/audioui/FakeAudioRepository.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+class FakeAudioOutputRepository(): AudioOutputRepository {
+    override val audioOutput: MutableStateFlow<AudioOutput> = MutableStateFlow(AudioOutput.None)
+    override val available: MutableStateFlow<List<AudioOutput>> = MutableStateFlow(listOf())
+
+    override fun launchOutputSelection(closeOnConnect: Boolean) {
+    }
+
+    override fun close() {
+    }
+}
```
[the remark this answers is not in the document]
Comment (kul3r4) [reply audio-ui/src/androidTest/java/com/google/android/horologist/audioui/FakeAudioRepository.kt]: done

File: audio-ui/src/androidTest/java/com/google/android/horologist/audioui/VolumeScreenTest.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import com.google.android.horologist.audio.ExperimentalAudioApi
+import com.google.android.horologist.audio.VolumeState
+import com.google.android.horologist.audioui.devices.RotaryInput
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.test.runTest
+import org.junit.Rule
+import org.junit.Test
+
+
+/*
```
[the remark this answers is not in the document]
Comment (kul3r4) [reply audio-ui/src/androidTest/java/com/google/android/horologist/audioui/VolumeScreenTest.kt]: done

File: audio-ui/src/main/java/com/google/android/horologist/audioui/EmulatedHaptics.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.horologist.audioui
+
+import com.google.android.horologist.audioui.devices.Haptics
+
+public object EmulatedHaptics : Haptics {
```
[the remark this answers is not in the document]
Comment (kul3r4) [reply audio-ui/src/main/java/com/google/android/horologist/audioui/EmulatedHaptics.kt]: removed

File: audio-ui/src/main/java/com/google/android/horologist/audioui/devices/Devices.kt (hunk 1 above)
Comment (kul3r4) [reply audio-ui/src/main/java/com/google/android/horologist/audioui/devices/Devices.kt]: removed

File: audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt [hunk 7]
```diff
@@ -64,4 +75,76 @@ public class VolumeScrollableState(private val volumeRepository: VolumeRepositor
...
+    }
+
+    private fun performStandardHaptics() {
+        perform(waveform)
+    }
+
+    private val composition = createComposition()
+
+    private val waveform = if (VERSION.SDK_INT >= VERSION_CODES.O) {
+        VibrationEffect.createWaveform(
```
[the remark this answers is not in the document]
Comment (kul3r4) [reply audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt:112]: I have added a TODO to make this a more appropriate waveform

File: audio-ui/src/androidTest/java/com/google/android/horologist/audioui/TestHaptics.kt (hunk 3 above)
Comment (kul3r4) [reply audio-ui/src/androidTest/java/com/google/android/horologist/audioui/TestHaptics.kt]: removed

File: audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (yschimke) [reply audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt:112]: What about for now, just simplifying this down to the one code path that works for GW4, and ignore haptics as a fallback (on API 25?).

There is a lot of code here, that I don't fully understand, that I'm not sure we know is as it should be, and we should get something smaller, tighter in.  then add this back this with appropriate review of the correct Haptics.

[the remark this answers is not in the document]
Comment (kul3r4) [reply audio-ui/src/main/java/com/google/android/horologist/audioui/VolumeScrollableState.kt]: Updated the code to only use an effect

**Pull request merged**